File: src/python/algorithms/backTracking.py
import numpy as np
import time
from algorithms.Core import *

def backTracking(m:int,n:int,k:int, datos,dataName:str):
    # Comenzamos a tomar el tiempo de ejecucion
    start = time.time()
	
    #creamos la grilla en x y en y para los datos 
    gridX:list = np.linspace(min(datos["x"]), max(datos["x"]), num=m, endpoint=True)
    gridY:list = np.linspace(min(datos["y"]), max(datos["y"]), num=n, endpoint=True)
    
    #sacamos el primer y ultimo elemento de la grilla x para trabajr unicamente con las columnas del medio
    conjunto_medio:list = gridX[1:-1]

    # creamos una lista que va a contener todas las combinaciones posibles de como separar las columnas del medio dado k-2
    listas_medio = []
    listasCombinatorias(conjunto_medio,[],listas_medio,k-2)

    #agregamos el primer y ultimo elemento a cada lista 
    listas_medio = [[gridX[0]] + lista + [gridX[-1]] for lista in listas_medio]

    bestRes = []
    bestSubGridX =  [] 
    bestError = 100000000000001

    # Agregamos unos valores por default necesarios para la ejecucion
    for i in range(k):
        bestRes.append(gridY[-1])
   
    #con cada subgrilla probamos fuerza bruta recursiva que recursivamente recorre todas las posibles combinaciones y devuelve el mejor error de esa subgrillaa
    for subGridX in listas_medio:
        print(subGridX)
        res = []
       
        #esto lo hice porque la lista estba vacia para darles valores inciales porque 
        #sino en la priemra iteracion no podia iterar sobre valoers vacios por eso daba el lsit index out of range [1]
        for i in range(len(subGridX)):
            res.append(gridY[0])
        
        errorActual = 100000000000001
        
        errorActual = backTrackingRecursiva(subGridX,gridY,[],[],res, errorActual, datos)
        
        #si otra subgrilla tiene un mejor error guarda esa 
        if(errorActual < bestError):
            bestSubGridX = subGridX.copy()
            bestError = errorActual
            bestRes = res.copy()    
    
    # Cortamos el timepo y guardamos el tiempo de ejecucion
    end = time.time()
    totalTime = (end - start) * 1000

    # Generamos el Grafico de la Instancia
    plot_puntos_y_linea(datos,bestSubGridX,bestRes,m,n,"BackTracking",bestError,totalTime,dataName)
    
    return np.round(bestError,decimals=2),bestRes,np.round(totalTime,decimals=2)



def backTrackingRecursiva(gridX,gridY,xs:list,ys:list,res:list,bestError,datos):
    # poda de optamilidad: si la rama que se esta calculando es mas grande que el mejor reusltado hasta ahora, descartala 
    errorActual = errorBreakPoints(xs,ys,datos)
    if(errorActual > bestError):
        return errorActual
    
    #Caso Base
    if(len(gridX) == len(xs)):
        
        if(errorActual < bestError):
            res.clear()
            res.extend(ys)
            bestError = errorActual

        return bestError
    
    #Caso recursivo
    i = len(xs)
    xs.append(gridX[i])
    currentBestError = bestError
    for j in gridY:
        ys.append(j)
        error = backTrackingRecursiva(gridX,gridY,xs,ys,res,currentBestError,datos)
        if(error < currentBestError):
            currentBestError = error
        ys.pop()
    xs.pop()

    return currentBestError

File: src/cpp/backTracking.cpp
#include "include/backTracking.hpp"

float backTrackingRecursiva(const vector<float> &gridX,
                            const vector<float> &gridY,
                            vector<float> &resX,
                            vector<float> &resY,
                            vector<float> &bestResY,
                            float &bestError,
                            const json &datos)
{

    // Poda de optimalidad
    if (resX.size() != 0)
    {
        float errorActual = errorBreakPoint(resX, resY, datos);
        if (errorActual > bestError)
        {
            return errorActual;
        }
    }

    // Caso Base
    if (gridX.size() == resX.size())
    {
        float errorBP = errorBreakPoint(resX, resY, datos);

        if (errorBP < bestError)
        {
            bestResY.clear();
            bestResY = resY;
            bestError = errorBP;
        }

        return bestError;
    }

    size_t i = resX.size();
    resX.push_back(gridX[i]);
    float currentBestError = bestError;

    // Caso recursivo
    for (const auto &j : gridY)
    {
        resY.push_back(j);
        float error = backTrackingRecursiva(gridX, gridY, resX, resY, bestResY, currentBestError, datos);
        if (error < currentBestError)
        {
            currentBestError = error;
        }
        resY.pop_back();
    }

    resX.pop_back();

    return currentBestError;
}

Result_t backTracking(int m, int n, int k, const json &data)
{
    // empezar el tiempo:
    auto inicio = high_resolution_clock::now();

    // creamos grilla de datos
    vector<float> gridX;
    vector<float> gridY;
    try
    {
        linspace(gridX, minJson(data, "x"), maxJson(data, "x"), m);
        linspace(gridY, minJson(data, "y"), maxJson(data, "y"), n);
    }
    catch (const std::invalid_argument &e)
    {
        exit(1);
    }

    printVector(&gridX, "GRID X: ");
    printVector(&gridY, "GRID Y: ");

    // sacamos el primer y ultimo elemento de la grilla x para trabajr unicamente con las columnas del medio
    vector<float> conjunto_medio = gridX;
    if (!conjunto_medio.empty())
    {
        conjunto_medio.erase(conjunto_medio.begin());
        conjunto_medio.pop_back();
    }

    // creamos una lista que va a contener todas las combinaciones posibles de como separar las columnas del medio dado k-2
    vector<vector<float>> listas_medio;
    vector<float> subconjunto;
    listasCombinatorias(conjunto_medio, subconjunto, listas_medio, k - 2);

    // agregamos el primer y ultimo elemento de la lista
    for (auto &lista : listas_medio)
    {
        lista.insert(lista.begin(), gridX[0]);
        lista.push_back(gridX.back());
    }

    vector<float> bestSubGridX;
    float bestError = 100000000000001.0f;
    vector<float> bestRes(k, gridY.back());

    // con cada subgrilla probamos backtracking recursiva que recursivamente recorre todas las posibles combinaciones y devuelve el mejor error de esa subgrillaa
    for (const auto &subGridX : listas_medio)
    {

        vector<float> res(subGridX.size(), gridY[0]);
        vector<float> resX;
        vector<float> resY;

        float errorActual = 100000000000001.0f;
        errorActual = backTrackingRecursiva(subGridX, gridY, resX, resY, res, bestError, data);

        if (errorActual < bestError)
        {
            bestSubGridX = subGridX;
            bestError = errorActual;
            bestRes = res;
        }
    }

    // calculamos el tiempo
    auto fin = high_resolution_clock::now();
    double totalTime = (double)duration_cast<duration<double>>(fin - inicio).count();

    // devolvemos un struct con todas las respuestas
    Result_t result = {bestError, bestSubGridX, bestRes, static_cast<float>(totalTime)};

    return result;
}
